src/main/scala/tech/cryptonomic/conseil/tezos/TezosRpc.scala

```diff
@@ -0,0 +1,189 @@
+    trait IOEff {
+      import tech.cryptonomic.conseil.tezos.TezosRpc.Akka.TezosNodeContext
+      import tech.cryptonomic.conseil.tezos.TezosRpc.Akka.Futures._
+      import tech.cryptonomic.conseil.util.EffectsUtil._
+
+      /* Creates an IO instance on top of the one for Futures
+       * We can't use an automatic functionK conversion, because
+       * the semantic of Future execution is eager, so we need to
+       * manually create a future in a lazy (by-name) parameter to
+       * prevent the future being run ahead-of-time
+       */
+      implicit def ioRpcHandlerInstance(
+        implicit context: TezosNodeContext
+      ): RpcHandler.Aux[IO, String, String, JsonString] =
+        new RpcHandler[IO, UrlPath, JsonContent] {
+          val futureRpc = futureRpcHandlerInstance
+
+          // payload is formally verified json
+          type PostPayload = JsonString
+
+          override def getQuery = Kleisli { in =>
+            toIO(futureRpc.getQuery.run(in))
+          }
+
+          override def postQuery = Kleisli {
+            case in => toIO(futureRpc.postQuery.run(in))
+          }
+        }
+
+    }
+
+  }
+
+  /** Instances based on Akka toolkit*/
+  object Akka {
+    import com.typesafe.scalalogging.Logger
+    import scala.concurrent.Future
+    import scala.util.control.NoStackTrace
+    import akka.actor.ActorSystem
+    import akka.stream.ActorMaterializer
+    import akka.http.scaladsl.Http
+    import akka.http.scaladsl.model._
+    import cats.ApplicativeError
+
+    /** A generic marker type for shutdown completion signals*/
+    trait ShutdownComplete
+
+    /** Describes configurations for the tezos node, needed to actually execute the calls */
+    case class TezosNodeContext(
+      tezosConfig: TezosConfiguration,
+      timeoutConfig: NetworkTimeoutConfiguration,
+      streamingConfig: HttpStreamingConfiguration
+    )(implicit val system: ActorSystem) {
+
+      object ShutdownComplete extends ShutdownComplete
+
+      private val rejectingCalls = new java.util.concurrent.atomic.AtomicBoolean(false)
+
+      def translateCommandToUrl(command: String): String = {
+        import tezosConfig.nodeConfig._
+        s"$protocol://$hostname:$port/${pathPrefix}chains/main/$command"
+      }
+
+      def shutdown(): Future[ShutdownComplete] = {
+        rejectingCalls.compareAndSet(false, true)
+        Http(system).shutdownAllConnectionPools().map(_ => this.ShutdownComplete)(system.dispatcher)
+      }
+
+      /** Allows to stop calling the service when a shutdown is initiated, short-circuiting the response */
+      def withRejectionControl[T, Eff[_], Err](
+        error: => Err
+      )(call: => Eff[T])(implicit errorEffect: ApplicativeError[Eff, Err]): Eff[T] =
+        if (rejectingCalls.get)
+          errorEffect.raiseError(error)
+        else
+          call
+
+    }
+
+    /** Statically provides rpc intances based on Akka and scala Future*/
+    object Futures extends Futures
```

Comment:
Could you explain why the trait and object `Futures` are needed here? Can't the implicit `RpcHandler` instances be directly placed in the `Akka` object?

Reply:
Having both trait and object allows one to import the definition in scope alternatively via
1. explicit import of the object `import TezosRpc.Akka.Futures._`
2. mixin of the trait `... extends TezosRpc.Akka.Futures`

I used the nested "module names" to keep the service backend separate from the typed-effect definition on which the implementation is based...
but on a second thought, it's quite far-fetched to think that we might even want to implement an Akka backend using anything else than Futures.